apis/python/src/tiledbsoma/dense_nd_array.py

```diff
@@ -148,18 +151,80 @@ def read_tensor(
+        must be less than or equal to the number of dimensions. For example, if the array
+        is 10 by 20, then some acceptable values of ``coords`` include ``(3, 4)``,
+        ``(slice(5, 10),)``, and ``(slice(5, 10), slice(6, 12))``. Slice indices are
+        doubly inclusive.
         """
-        tiledb_result_order = tiledb_result_order_from_soma_result_order(
-            result_order, accept=["column-major", "row-major"]
-        )
         with self._tiledb_open("r") as A:
             target_shape = dense_indices_to_shape(coords, A.shape, result_order)
-            query = A.query(return_arrow=True, order=tiledb_result_order)
-            arrow_tbl = query.df[coords]
-            return pa.Tensor.from_numpy(
-                arrow_tbl.column("soma_data").to_numpy().reshape(target_shape)
+            schema = A.schema
+            ned = A.nonempty_domain()
+
+        sr = clib.SOMAReader(
+            self._uri,
+            name=self.__class__.__name__,
+            result_order=result_order,
+            platform_config={} if self._ctx is None else self._ctx.config().dict(),
+        )
+
+        if coords is not None:
+            if not isinstance(coords, (list, tuple)):
+                raise TypeError(
+                    f"coords type {type(coords)} unsupported; expected list or tuple"
+                )
+            if len(coords) < 1 or len(coords) > schema.domain.ndim:
+                raise ValueError(
+                    f"coords {coords} must have length between 1 and ndim ({schema.domain.ndim}); got {len(coords)}"
+                )
+
+            for i, coord in enumerate(coords):
+                dim_name = schema.domain.dim(i).name
+                if coord is None:
+                    pass  # No constraint; select all in this dimension
+                elif isinstance(coord, int):
+                    sr.set_dim_points(dim_name, [coord])
+                elif isinstance(coord, slice):
+                    lo_hi = util.slice_to_range(coord, ned[i]) if ned else None
+                    if lo_hi is not None:
+                        lo, hi = lo_hi
+                        if lo < 0 or hi < 0:
+                            raise ValueError(
+                                f"slice start and stop may not be negative; got ({lo}, {hi})"
+                            )
+                        if lo > hi:
+                            raise ValueError(
+                                f"slice start must be <= slice stop; got ({lo}, {hi})"
+                            )
+                        sr.set_dim_ranges(dim_name, [lo_hi])
+                    # Else, no constraint in this slot. This is `slice(None)` which is like
+                    # Python indexing syntax `[:]`.
+                else:
+                    raise TypeError(f"coord type {type(coord)} at slot {i} unsupported")
+
+        sr.submit()
+
+        arrow_tables = []
+        while arrow_table_piece := sr.read_next():
```

Comment:
I believe the walrus operator is Python 3.8+.  I was under the impression SOMA was supporting 3.7+?

Site note: need a GH Action to validate the full test matrix at some point.

Reply:
> Site note: need a GH Action to validate the full test matrix at some point.

@bkmartinjr can you please elaborate?

Reply:
We should choose & and enforce it.  I was unaware of Literal :blush:

If it is 3.8+, I"m fine with walrus.

Reply:
@bkmartinjr I prefer to let #544 do its assigned duties